internal/envoy/v3/listener.go

```diff
@@ -460,52 +460,72 @@ func TCPProxy(statPrefix string, proxy *dag.TCPProxy, accesslogger []*accesslog.
 	// The value of two and a half hours for reasons documented at
 	// https://github.com/projectcontour/contour/issues/1074
 	// Set to 9001 because now it's OVER NINE THOUSAND.
-	idleTimeout := protobuf.Duration(9001 * time.Second)
+	tcpProxy := &tcp.TcpProxy{
+		StatPrefix:  statPrefix,
+		AccessLog:   accesslogger,
+		IdleTimeout: protobuf.Duration(9001 * time.Second),
+	}
+
+	// Get the total weight of all clusters for the proxy.
+	var totalWeight uint32
+	for _, c := range proxy.Clusters {
+		totalWeight += c.Weight
+	}
+
+	// If some clusters have non-zero weights, then drop
+	// any others that have zero weights since they shouldn't
+	// get traffic (note that TCPProxy weighted clusters can't
+	// have zero weights, unlike HTTP route weighted clusters,
+	// so we can't include them with a zero weight). Also note
+	// that if all clusters have zero weights, then we keep them
+	// all and evenly weight them below.
+	var keepClusters []*dag.Cluster
+	for _, c := range proxy.Clusters {
+		if totalWeight > 0 && c.Weight == 0 {
```

Comment:
just a thought how to do this in one pass (havent tested it tho)

```
var totalWeight uint32
var weightedClusters []*dag.Cluster
for _, c := range proxy.Clusters {
  if c.Weight != 0 {
    weightedClusters = append(weightedClusters, c)
  }
  totalWeight += c.Weight
}

var actualClusters []*dag.Cluster
if totalWeight == 0 {
  actualClusters = proxy.Clusters
} else {
  actualClusters = weightedClusters
}

switch len(actualClusters) {
...
```

could rename `weightedClusters` to `clustersWithWeight` too

Reply:
:+1: I prioritized readability over perf here, but made an update here along the lines of ^^^ that I think is still easy to read/follow.

Reply:
nice! yeah good adaptation